federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,230 @@ limitations under the License.
-func getFederationDNSZoneName() string {
-	return "mydomain.com" // TODO: quinton: Get this from the federation configuration.
+func (s *ServiceController) getFederationDNSZoneName() (string, error) {
+	return "example.com", nil // TODO: quinton: Get this from the federation configuration.
+	// Note: For unit testing this must match the domain populated in the test/stub dnsprovider.
+}
+
+// getZone is a hack around the fact that dnsprovider does not yet support a Get() method, only a List() method.  TODO:  Fix that.
+func getZone(dnsZoneName string, zonesInterface dnsprovider.Zones) (dnsprovider.Zone, error) {
+	zones, err := zonesInterface.List()
+	if err != nil {
+		return nil, err
+	}
+	for _, zone := range zones {
+		if zone.Name() == dnsZoneName {
+			return zone, nil
+		}
+	}
+	return nil, fmt.Errorf("DNS zone %s not found.", dnsZoneName)
 }
 
-func ensureDNSRecords(clusterName string, cachedService *cachedService) error {
+// getRrset is a hack around the fact that dnsprovider.ResourceRecordSets interface does not yet include a Get() method, only a List() method.  TODO:  Fix that.
+func getRrset(dnsName string, rrsetsInterface dnsprovider.ResourceRecordSets) (dnsprovider.ResourceRecordSet, error) {
+	rrsets, err := rrsetsInterface.List()
+	if err != nil {
+		return nil, err
+	}
+	for _, rrset := range rrsets {
+		if rrset.Name() == dnsName {
+			return rrset, nil
+		}
+	}
+	return nil, fmt.Errorf("Rrset %s not found.", dnsName)
+}
+
+/* ensureDnsRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resourcerecordsets for dnsName are consistent with endpoints.
+   if endpoints is nil or empty, a CNAME record to uplevelCname is ensured
+*/
+func (s *ServiceController) ensureDnsRrsets(dnsZoneName, dnsName string, endpoints []string, uplevelCname string) error {
+	zone, err := getZone(dnsZoneName, s.zones) // TODO: s.zones.Get(dnsZoneName)
+	if err != nil {
+		return err
+	}
+	rrsets, supported := zone.ResourceRecordSets()
+	if !supported {
+		return fmt.Errorf("Failed to ensure DNS records for %s. DNS provider does not support the ResourceRecordSets interface.", dnsName)
+	}
+	rrset, err := getRrset(dnsName, rrsets) // TODO: rrsets.Get(dnsName) // TODO: quinton - need more params here.
+	if err != nil {
+		// Assume that it doesn't exist yet, so create it
+		if endpoints == nil || len(endpoints) < 1 {
+			// There are no endpoint addresses at this level, so CNAME to uplevel, if provided
+			if uplevelCname != "" {
+				newRrset := rrsets.New(dnsName, []string{uplevelCname}, minDnsTtl, rrstype.CNAME)
+				rrset, err = rrsets.Add(newRrset)
+				if err != nil {
+					return err
+				}
+			}
+			// We want no record, and we have no record, so we're all good.
+		} else {
+			// We have valid endpoint addresses, so just add them as A records.
+			newRrset := rrsets.New(dnsName, endpoints, minDnsTtl, rrstype.A)
+			// TODO: quinton ^^ - for AWS, an endpoint might be a hostname, so use a CNAME instead...
```

Comment:
OK, I've addressed this by resolving all DNS names to IP addresses, and adding those to the A record.
It's not ideal, in that we'll have to keep doing that over time to keep track of potential DNS changes by cloud providers,
but there's not way around this as far as I can see.  In an ideal world we would have a mixture of A records and CNAME records associated with this name, and allow clients to choose any one of them, but the DNS spec does not allow that.